### PR TITLE
Set node by name

### PR DESCRIPTION
**Pull request by ollmer**
1. Use names instead of indices in the `SetNextNode` and everywhere.
2. Remove node counter logic from the tape view. Compute the next intended node in the agent.
3. Track the last used node and the last SetNextNode() step in the view
4. Update all tapes accordingly
5. Rewrite the node intro in the intro.ipynb

### CODE REVIEW

**Comment by ollmer** [comment]
@rizar pls pay special attention to the view.py (I've changed it since our last talk, it property instead of methods again) and intro_clean.ipynb (I've rewritten the nodes intro to reflect addressing by name)

File: tapeagents/prompting.py [hunk 1]
```diff
@@ -61,7 +61,8 @@ def prompt_with_guidance(tape: Tape, guidance: str) -> Prompt:
 def view_to_messages(view: TapeView, agent: Agent | None = None) -> list[dict]:
     messages = []
     for step in view.steps:
-        assert not isinstance(step, (Pass, SetNextNode))
+        if isinstance(step, (Pass, SetNextNode)):
```
Comment (rizar) [inline tapeagents/prompting.py:64]: before `Pass` and `SetNextNode` would not make it into the view, I suppose you changed that

File: tapeagents/view.py [hunk 2]
```diff
@@ -151,9 +150,9 @@ def put_new_view_on_stack(self, step):
         self.messages_by_agent[receiver].append(step)
 
     @staticmethod
-    def compute(tape: Tape) -> TapeViewStack[StepType]:
+    def compute(tape: Tape, agent_name: str = "root") -> TapeViewStack[StepType]:
```
Comment (rizar) [inline tapeagents/view.py:153]: I'd like to be slightly more explicit here

```suggestion
    def compute(tape: Tape, root_agent_name: str = "root") -> TapeViewStack[StepType]:
```

File: examples/intro_clean.ipynb [hunk 3]
```diff
@@ -1124,22 +1145,22 @@
     "        )\n",
     "\n",
     "    def generate_steps(self, agent, dialog, llm_stream: LLMStream):\n",
-    "        o = llm_stream.get_output()\n",
-    "        if o.content:\n",
-    "            yield SetNextNode(next_node=0)\n",
-    "            yield AssistantStep(content=o.content)\n",
-    "        elif o.tool_calls:\n",
-    "            yield SetNextNode(next_node=1)\n",
+    "        m = llm_stream.get_output()\n",
```
Comment (rizar) [inline examples/intro_clean.ipynb:1148]: if we call it "output", we should probably use "o" as the abbreviation and not "m"?

File: examples/intro_clean.ipynb [hunk 4]
```diff
@@ -1026,24 +1040,27 @@
...
-    "            yield Respond(content=o.content)\n",
-    "        elif o.tool_calls:\n",
-    "            # while the LLM suggests tool calls, yield them as action steps\n",
-    "            yield ToolCalls.from_llm_output(o)\n",
-    "            yield SetNextNode(next_node=0)\n",
+    "            yield Respond(content=m.content)\n",
+    "        elif m.tool_calls:\n",
+    "            # when the LLM suggests tool calls, yield them as action steps\n",
+    "            yield ToolCalls.from_llm_output(m)\n",
+    "            print(f\"{agent.full_name}: search, yield main\")\n",
```
Comment (rizar) [inline examples/intro_clean.ipynb:1062]: it this forgotten debug print?

**Review by rizar — APPROVED** [review]
this looks good! 

can you please confirm that now control flow steps will remain in the view and should be removed and the time of prompt creation?

File: tapeagents/prompting.py (hunk 1 above)
Comment (ollmer) [reply tapeagents/prompting.py:64]: We've discussed this during the call, leaving it as it is.

File: examples/intro_clean.ipynb (hunk 3 above)
Comment (ollmer) [reply examples/intro_clean.ipynb:1148]: Yep, artifact of the main branch merge, will fix

**Pull request merged**